pkg/resources/kafka/kafka.go

```diff
@@ -297,44 +297,75 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 
 func (r *Reconciler) reconcileKafkaPodDelete(log logr.Logger) error {
 	podList := &corev1.PodList{}
-	matchingLabels := client.MatchingLabels(kafka.LabelsForKafka(r.KafkaCluster.Name))
 
-	err := r.Client.List(context.TODO(), podList, client.ListOption(client.InNamespace(r.KafkaCluster.Namespace)), client.ListOption(matchingLabels))
+	err := r.Client.List(context.TODO(), podList,
+		client.ListOption(client.InNamespace(r.KafkaCluster.Namespace)),
+		client.ListOption(client.MatchingLabels(kafka.LabelsForKafka(r.KafkaCluster.Name))),
+		client.ListOption(client.HasLabels{"brokerId"}),
+	)
 	if err != nil {
 		return errors.WrapIf(err, "failed to reconcile resource")
 	}
 
-	deletedBrokers := make([]corev1.Pod, 0)
-OUTERLOOP:
+	brokerIDsFromSpec := make(map[string]bool, len(r.KafkaCluster.Spec.Brokers))
+	for _, broker := range r.KafkaCluster.Spec.Brokers {
+		brokerIDsFromSpec[strconv.Itoa(int(broker.Id))] = true
+	}
+
+	podsDeletedFromSpec := make([]corev1.Pod, 0, len(podList.Items))
```

Comment:
The reason for using the `len(podList.Items)` for setting the capacity is that that is the maximum number of broker pods which can be removed from the cluster and to avoid unnecessary allocations. I personally don't see issue with the current implementation.

Reply:
understandable